x/smartaccount/ante/smartaccount_auth.go

```diff
@@ -0,0 +1,98 @@
+package ante
+
+import (
+	"encoding/json"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/terra-money/core/v2/x/smartaccount/types"
+)
+
+// SmartAccountAuthDecorator does authentication for smart accounts
+type SmartAccountAuthDecorator struct {
+	sak                       SmartAccountKeeper
+	wk                        WasmKeeper
+	ak                        authante.AccountKeeper
+	defaultVerifySigDecorator sdk.AnteHandler
+}
+
+func NewSmartAccountAuthDecorator(
+	sak SmartAccountKeeper,
+	wk WasmKeeper,
+	ak authante.AccountKeeper,
+	sigGasConsumer func(meter sdk.GasMeter, sig signing.SignatureV2, params authtypes.Params) error,
+	signModeHandler authsigning.SignModeHandler,
+) SmartAccountAuthDecorator {
+	if sigGasConsumer == nil {
+		sigGasConsumer = authante.DefaultSigVerificationGasConsumer
+	}
+	defaultVerifySigDecorator := sdk.ChainAnteDecorators(
+		authante.NewSetPubKeyDecorator(ak),
+		authante.NewValidateSigCountDecorator(ak),
+		authante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
+		authante.NewSigVerificationDecorator(ak, signModeHandler),
+	)
+	return SmartAccountAuthDecorator{
+		sak:                       sak,
+		wk:                        wk,
+		defaultVerifySigDecorator: defaultVerifySigDecorator,
+	}
+}
+
+// AnteHandle checks if the tx provides sufficient fee to cover the required fee from the fee market.
+func (sad SmartAccountAuthDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// check if the tx is from a smart account
+	setting, err := sad.sak.GetSetting(ctx, tx.GetMsgs()[0].GetSigners()[0].String())
+	if sdkerrors.ErrKeyNotFound.Is(err) {
+		// run through the default handlers for signature verification
+		newCtx, err := sad.defaultVerifySigDecorator(ctx, tx, simulate)
+		if err != nil {
+			return newCtx, err
+		}
+		// continue to the next handler after default signature verification
+		return next(newCtx, tx, simulate)
+	} else if err != nil {
+		return ctx, err
+	}
+
+	if setting.Authorization != nil && len(setting.Authorization) > 0 {
```

Comment:
@freeelancer should also handle `simulate` case so simulation doesnt require any authorization